Profiling

seemk opened the pull request:
# Description

Add support for CPU profiling.

Brings in new environment variables:
* `SPLUNK_PROFILER_ENABLED` - disabled by default
* `SPLUNK_PROFILER_LOGS_ENDPOINT` - defaults to `http://localhost:4317`
* `SPLUNK_PROFILER_CALL_STACK_INTERVAL` - thread sampling (stacktrace taking) interval in milliseconds, defaults to `1000`
* `SPLUNK_PROFILER_INCLUDE_INTERNAL_STACKS` - whether to include stacktraces from internal profiler threads (sampler thread and logs processor/exporter), disabled by default.

The profiler / sampling works from a different thread, where call stacks for all threads are periodically sampled. These are then serialized to protobuf (pprof format), gzipped and encoded as base64. The base64 content are sent to collector as OTel logs. For the exporting we can leverage the OTel logs SDK, which makes it quite convenient.

To correlate active span contexts with sampled stacktraces, `opentelemetry.context.{attach,detach}` are monkeypatched, to store a mapping of `thread ID -> span context`. This mapping is available for the sampler thread.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Tested manually
- [x] Added automated tests

# Checklist:

- [ ] Changelogs have been updated
- [x] Unit tests have been added/updated
- [ ] Documentation has been updated

nstawski commented:
@Mergifyio rebase

nstawski commented:
@seemk please rebase this pull against `main`

nstawski commented:
@seemk also the `lint` job is failing.

tsloughter-splunk commented:
Is there a design doc or can you describe how it is working?

Not knowing much about this the one thing that caught my eye when trying to review is the global batch log processor, but I assume that is the best solution? :)

But do other profilers use their own log processor/exporters instead of using one that might already be created by the application for logs?

seemk commented:
> Is there a design doc or can you describe how it is working?
> 
> Not knowing much about this the one thing that caught my eye when trying to review is the global batch log processor, but I assume that is the best solution? :)
> 
> But do other profilers use their own log processor/exporters instead of using one that might already be created by the application for logs?

We went over the workings during a call, but I'll write a summary into the PR description.

As for the global batch processor, I just used it for convenience to get the export pipeline set up. Other distros (Node) didn't have the logs SDK available when profiling was implemented and had to have a custom exporter for the logs protos.

pmcollins commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,300 @@
+    return gzip.compress(pb_profile.SerializeToString())
+
+
+def _profiler_loop(options: _Options):
```
What do you think about making this functionality unit-testable?

seemk replied:
Split this function into multiple smaller ones now, I think it is currently implicitly covered by our unit tests

pmcollins reviewed (commented):
Thanks for doing this Siim. A few questions/comments below.

seemk commented on tests/unit/test_profiling.py:
```diff
@@ -0,0 +1,169 @@
+        return False
+
+    def test_profiling_export(self):
+        provider = TracerProvider()
```
In this case the goal is to verify that the spans we generate for the `do_work` function are correlated with the profiling stacktraces. In `_assert_log_record` the profiling data and the active span context are checked for equality.

If tracing isn't set up, then profiling will work, but span context correlation will not.

pmcollins replied:
Ah, thank you. Can we capture that in the docs for this function?

seemk replied:
Added comments to the profiling test setup

seemk commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,300 @@
+import base64
+import gzip
+import logging
+import os
```
Indeed, good catch. This passes the linter :thinking: I''ll check if there's some lint config issue

seemk replied:
Got rid of the extra imports, but I couldn't figure out what's wrong with the linter. Tried upgrading to the latest pylint, but that had an unrelated error so I decided not to spend more time on it for now.

seemk replied:
I've started to use pyright locally now and it catches the unused imports.

Created https://github.com/signalfx/splunk-otel-python/issues/354

Perhaps worth switching to pyright in the future?

seemk commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,300 @@
+logger = logging.getLogger(__name__)
+
+
+class Profiler:
```
In this case I wanted to just store the globals somewhere (batch processor, running flag), not sure hiding behind these methods will be more useful, or will it?

pmcollins replied:
I do think encapsulating a class's fields behind methods makes the class easier to understand and maintain because the amount of API exposed becomes smaller, and then callers don't have to know about the class's internals.

seemk commented on splunk_otel/distro.py:
```diff
@@ -47,8 +48,14 @@ def __init__(self):
+            _configure_tracing(options)
+
+        if self._profiling_enabled:
+            _start_profiling(ProfilingOptions(options.resource))
```
The issue stemmed from wanting to only do the resource detection once in our distro (`_SplunkDistro._configure`), so I added 2 different `start_profiling` functions: `_start_profiling(Options)` which is supposed to be used internally in `distro.py` and `start_profiling(service_name, resource_attributes, endpoint)` to be used with manual setup and to retain parity with `start_tracing`. In the tracing case it is `start_tracing` for manual instrumentation and `_configure_tracing` used in the distro.

So in this case the users would be able to use [manual instrumentation](https://docs.splunk.com/observability/en/gdi/get-data-in/application/python/instrumentation/python-manual-instrumentation.html) like:
```python
from splunk_otel.tracing import start_tracing

service_name = "foo"
resource = { "xyz": 42 }
start_tracing(service_name, resource)
start_profiling(service_name, resource)
```

The underlying issue is that in the end we'll need a single `start` function, but I think this is work better kept out from the profiling PR. In Node we initially had multiple `start_*` functions, which then got a deprecation marker once an unified `start` became available.

seemk replied:
Perhaps the better way would be to move the configuration defaults to a single file (i.e. the static functions from both tracing and profiling options) and keep the Options classes as only containers for the variables, but I'd keep it out of this PR. I do think the tracing and profiling options need to be separate though.

seemk commented on splunk_otel/profiling/__init__.py:
I think it would make sense to address it in the same work where `start` functions will be unified for the manual API.

seemk commented:
> Is there a design doc or can you describe how it is working?
> 
> Not knowing much about this the one thing that caught my eye when trying to review is the global batch log processor, but I assume that is the best solution? :)
> 
> But do other profilers use their own log processor/exporters instead of using one that might already be created by the application for logs?

Added a small description how the "profiler" operates, it turned out to be pretty small in Python.

seemk commented:
Added filtering of internal threads (enabled by default) and `stop_profiling`

seemk commented:
* Added reinitialization of the profiler thread on forks (for forking servers such as gunicorn).
* Added `LoggerProvider` instead of working with the batch processor directly. The logger provider has exit handlers, where it will shut down the exporting pipeline.

pmcollins reviewed (commented):
I'm wondering if we can modify how we access the `Profiler` global variable, such that its global state is accessed once by a global function, but all other access is via function arguments.

seemk commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,300 @@
+    service_name: Optional[str] = None,
+    resource_attributes: Optional[Dict[str, Union[str, bool, int, float]]] = None,
+    endpoint: Optional[str] = None,
+    call_stack_interval: Optional[int] = None,
```
Renamed it to `call_stack_interval_millis`

seemk commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,300 @@
+    call_stack_interval: Optional[int] = None,
+):
+    # pylint: disable-next=protected-access
+    resource = splunk_otel.options._Options._get_resource(
```
Yeah, see my comment regarding moving the default configuration functions to a separate location, some of them (such as resource generation) are shared, but some of them would be not. I don't want to modify the tracing options with this PR.

seemk commented on splunk_otel/profiling/__init__.py:
```diff
@@ -0,0 +1,405 @@
+        logger.warning("profiler already running")
+        return
+
+    _profiler.options = options
```
If `Profiler` was an user-facing API, why not. But in this case it's only a private raw state holder (to refrain from using multiple globals and having to use `global X` when doing assignments). I think adding these functions would only cause unnecessary indirection when reading code, at the moment I don't see a reason to do it. If there's a good reason, sure - I might be missing something as well.

seemk replied:
Regarding the constructor: I want the default state to be uninitialized state, so I don't have to check for the existence of the profiler (e.g. `if _profiler and _profiler._logger_provider`) in each of the global-accessing functions (`stop_profiling`, `force_flush`, `start_profiling`). Could add a `start(options)` functions which initializes it completely, or do you mean making `options` optional in the constructor and initialize the remainder only if they are passed in?

pmcollins replied:
A couple of principles that I try to remember to use are:

* Minimize access to global variables. If you must access a global variable, do it once in a getter (and maybe once in a setter). Other than that, pass it around as a function argument (like [this code](https://github.com/signalfx/splunk-otel-python/pull/327/files#diff-4d67cd7b39484835d62dd8b53be7673692c73194cb05af6704eb6df534981445R257)). (Then almost all of your tests won't have to prepare/manipulate global state either.)
* With the exception of data classes, don't access a class's fields from outside the class. Instead encapsulate field access using methods.

A lot of the functions in this file are exposed as functions that modify global state, but look to me like they would be cohesive to Profiler as methods (because they read and write Profiler fields and perform related logic).

Regarding your last question above -- I don't really have a strong opinion, although I would prefer that the `Profiler` not be coupled to `_Options` at all and just be able to be initialized in a constructor or, if necessary, some other method(s) that sets its state after construction.

seemk replied:
Removed access to the `_profiler` global from `_start_profiler_thread` and `_encode_cpu_profile` and made `_wrapped_context_{attach,detach}` use a local profiler given to them via `_make_wrapped_context_{attach,detach}`.

The remainder of the functions having access to it are `_start_profiling`, `_force_flush`, `stop_profiling`.

They could have an optional profiler argument, defaulting to the global profiler, but why if this codepath is never used? If it is eventually needed for unit testing, it can be added, otherwise it's just dead code.

I'm not sure what a global `get_profiler` would get us over accessing `_profiler` directly. There is no locking involved and it would not make a difference when mocking. Especially if the function would look like:

```python
def get_profiler():
  return _profiler
```

seemk replied:
>A lot of the functions in this file are exposed as functions that modify global state, but look to me like they would be cohesive to Profiler as methods (because they read and write Profiler fields and perform related logic).

Which ones do you have in mind?

pmcollins replied:
Hey Siim, I have created a new PR against this PR to demonstrate what I'm proposing: https://github.com/signalfx/splunk-otel-python/pull/355

seemk replied:
Thanks, I don't mind, just some of the changes (such as `get_profiler`) make me question what do we gain from using a function when it doesn't do anything, but it definitely isn't worth the time to keep arguing. Can you mark it as ready? I can merge it then

pmcollins replied:
Feel free to remove the function -- I don't think it's that important.

pmcollins commented:
I have read the CLA Document and I hereby sign the CLA

pmcollins commented:
Looks great, thanks for taking my feedback.

seemk commented:
@pmcollins Any outstanding issues you want addressed?

pmcollins commented:
> @pmcollins Any outstanding issues you want addressed?

LGTM. Unfortunately I don't have approval privileges though so you'll have to get that from someone else unless I can get privileges beforehand.
